### PR TITLE
[WIP] New metadata system

### PR DESCRIPTION
**Pull request by dunglas**
**Work in progress, not finished, not tested**

Goals:

- Ease resources and properties configuration (use annotations, no more service definition required)
- Ability to define easily custom operations (set the Symfony route to use in the annotation and you're all set)
- Operations can have different serialization/validation/whatever groups and options (the new `attributes` array will be used for that)
- Metadata are immutable value objects
- No hard coupling to Symfony anymore (next steps: Laravel and Silex support)
- Use the Symfony PropertyInfo component instead of `dunglas/property-info`
- Move all classes adding support of external libraries (Symfony, Doctrine) to the `Bridge` namespace (next steps: Pomm and Ting support)
- Use the decorator pattern and the new `decorate-priority` option of the Symfony DIC instead of the `ChainLoader`, tags and compiler pass (better class responsibility, easier to extend, less code so easier maintenance)

Task list:
- [x] Service definitions and priorities
- [x] Unit tests
- [x] Use the new metadata system in the routing system
- [x] Make Behat green
- [ ] HHVM support

### CODE REVIEW

File: Api/ResourceClassResolver.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        if (is_object($value)) {
+            $typeToFind = $type = $this->getObjectClass($value);
+        } elseif (null === $resourceClass) {
+            throw new InvalidArgumentException(sprintf('No resource class found.'));
+        } else {
+            $typeToFind = $resourceClass;
+        }
+
+        if (!$this->isResourceClass($typeToFind) || ($strict && isset($type) && $resourceClass !== $type)) {
+            if (is_subclass_of($type, $resourceClass)) {
```
Comment (sstok) [inline Api/ResourceClassResolver.php]: $type can be unset when only $resourceClass is set and isResourceClass() returns false.

File: Bridge/Symfony/Routing/ApiLoader.php [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+        if (isset($attributes['route_name'])) {
+            return;
+        }
+
+        if (!isset($attributes['method'])) {
+            throw new RuntimeException('Either a "route_name" or a "method" operation attribute must exist.');
+        }
+
+        if (isset($attributes['controller'])) {
```
Comment (sstok) [inline Bridge/Symfony/Routing/ApiLoader.php]: I guess this needs to be a negative? Because with the `else` you would get an undefined key notice.

**Comment by dunglas** [comment]
Waiting for the merge of symfony/symfony#16947

**Comment by dunglas** [comment]
Behat tests are green, this is ready to be tested. /cc @sroze

**Comment by sroze** [comment]
@dunglas can you squash the commits before 31c06a5? So I can PR stuff and we can keep the commit history? :wink:

**Comment by dunglas** [comment]
@sroze done (and rebased against master).

Any help to fix unit tests is welcome!

**Comment by sroze** [comment]
@dunglas looks like even AppVeyor's PR's passed, I'm in favour of merging this PR now. I'm globally happy with what I've seen, and as it's just impossible to review properly at the moment because of its size, we can continue the discussions over issues and PRs.

**Comment by teohhanhui** [comment]
Does this mean those of us who are using the `master` branch are now stranded? (Symfony 3.1 is still a few months away...)

**Comment by dunglas** [comment]
@teohhanhui not really, `master` requires a few SF 3.1 components for now (basically the serializer). Other Symfony components use the stable version (3.0 or even 2.8).

**Comment by teohhanhui** [comment]
I know this is out of scope, but I've always found it hard to mix and match different versions of Symfony components when using the full framework (the overrides are hard to get right). Anyway, I'll give it a try. :smile:

**Comment by dunglas** [comment]
you can do something like that in `composer.json`:

```
"require": { "symfony/symfony": "~3.0", "symfony/serializer": "~3.1@dev" }
```

**Comment by 3kynox** [comment]
How to expose resources now ? As it is done via annotations ? Any sample ?

**Comment by dunglas** [comment]
@3kynox https://github.com/dunglas/DunglasApiBundle/blob/master/tests/Fixtures/TestBundle/Entity/Dummy.php

You just need to mark your entity class with the `@Resource` annotation.

**Comment by 3kynox** [comment]
Thanks !

**Pull request merged**